### PR TITLE
[System]: Correctly implement close and shutdown in SslStream.

### PR DESCRIPTION
**Pull request by baulig**

### CODE REVIEW

File: mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs [hunk 1]
```diff
@@ -78,51 +85,56 @@ abstract class MobileAuthenticatedStream : AuthenticatedStream, MSI.IMonoSslStre
...
-		}
-
-		internal void CheckThrow (bool authSuccessCheck)
+		internal void CheckThrow (bool authSuccessCheck, bool shutdownCheck = false)
 		{
-			if (closeRequested != 0)
-				throw new InvalidOperationException ("Stream is closed.");
 			if (lastException != null)
-				throw lastException;
+				throw lastException.SourceException;
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs]: Use lastException.Throw () instead

File: mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs [hunk 2]
```diff
@@ -161,218 +171,219 @@ public IAsyncResult BeginAuthenticateAsServer (X509Certificate serverCertificate
...
+					throw new ArgumentException (nameof (serverCertificate));
+			} else {
+				if (targetHost == null)
+					throw new ArgumentException (nameof (targetHost));
+				if (targetHost.Length == 0)
+					targetHost = "?" + Interlocked.Increment (ref uniqueNameInteger).ToString (NumberFormatInfo.InvariantInfo);
+			}
+
+			if (lastException != null)
+				throw lastException.SourceException;
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs]: lastException.Throw ()

File: mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs [hunk 3]
```diff
@@ -161,218 +171,219 @@ public IAsyncResult BeginAuthenticateAsServer (X509Certificate serverCertificate
...
-			EndReadOrWrite (asyncResult, ref asyncWriteRequest);
+			TaskToApm.End (asyncResult);
 		}
 
 		public override int Read (byte[] buffer, int offset, int count)
 		{
-			return ProcessReadOrWrite (ref asyncReadRequest, ref readBuffer, ProcessRead, new BufferOffsetSize (buffer, offset, count), null);
+			var asyncRequest = new AsyncReadRequest (this, true, buffer, offset, count);
+			var task = StartOperation (OperationType.Read, asyncRequest, CancellationToken.None);
+			task.Wait ();
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs]: This is redundant, .Result does that

File: mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs [hunk 4]
```diff
@@ -161,218 +171,219 @@ public IAsyncResult BeginAuthenticateAsServer (X509Certificate serverCertificate
...
-			Debug ("ProcessReadOrWrite: {0} {1}", name, userBuffer);
+			CheckThrow (true, type != OperationType.Read);
+			Debug ("StartOperationAsync: {0} {1}", asyncRequest, type);
 
-			var asyncRequest = new AsyncProtocolRequest (this, lazyResult, userBuffer);
-			return StartOperation (ref nestedRequest, ref internalBuffer, operation, asyncRequest, name);
-		}
+			if (type == OperationType.Read) {
+				if (Interlocked.CompareExchange (ref asyncReadRequest, asyncRequest, null) != null) {
+					Console.Error.WriteLine ("INVALID NESTED CALL!");
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs]: Remove the Console.WriteLine

File: mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs [hunk 5]
```diff
@@ -161,218 +171,219 @@ public IAsyncResult BeginAuthenticateAsServer (X509Certificate serverCertificate
...
+					throw new InvalidOperationException ("Invalid nested call.");
+				}
+			} else {
+				if (Interlocked.CompareExchange (ref asyncWriteRequest, asyncRequest, null) != null) {
+					Console.Error.WriteLine ("INVALID NESTED CALL!");
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MobileAuthenticatedStream.cs]: Remove the Console.WriteLine

File: mcs/class/System/Mono.Btls/MonoBtlsObject.cs [hunk 6]
```diff
@@ -102,6 +102,20 @@ protected void CheckError (int ret, [CallerMemberName] string callerName = null)
 			CheckError (ret == 1, callerName);
 		}
 
+		protected internal void CheckLastError ([CallerMemberName] string callerName = null)
```
Comment (marek-safar) [inline mcs/class/System/Mono.Btls/MonoBtlsObject.cs]: Does it have to here when it's called from one place only?
Could also be make static

File: mcs/class/System/Mono.Btls/MonoBtlsObject.cs [hunk 7]
```diff
@@ -102,6 +102,20 @@ protected void CheckError (int ret, [CallerMemberName] string callerName = null)
...
+		{
+			var error = Interlocked.Exchange (ref lastError, null);
+			if (error == null)
+				return;
+
+			string message;
+			if (callerName != null)
+				message = string.Format ("Caught unhandled exception in {0}.{1}.", GetType ().Name, callerName);
+			else
+				message = string.Format ("Caught unhandled exception.");
```
Comment (marek-safar) [inline mcs/class/System/Mono.Btls/MonoBtlsObject.cs]: unused string.Format

File: mcs/class/System/Mono.Btls/MonoBtlsObject.cs [hunk 8]
```diff
@@ -102,6 +102,20 @@ protected void CheckError (int ret, [CallerMemberName] string callerName = null)
...
 
+		protected internal void CheckLastError ([CallerMemberName] string callerName = null)
+		{
+			var error = Interlocked.Exchange (ref lastError, null);
+			if (error == null)
+				return;
+
+			string message;
+			if (callerName != null)
+				message = string.Format ("Caught unhandled exception in {0}.{1}.", GetType ().Name, callerName);
```
Comment (marek-safar) [inline mcs/class/System/Mono.Btls/MonoBtlsObject.cs]: Use $"" syntax

**Comment by baulig** [comment]
build

**Pull request merged**